airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/query.py

```diff
@@ -0,0 +1,1387 @@
+                            node {
+                                __typename
+                                id
+                                namespace
+                                value
+                                key
+                                description
+                                createdAt
+                                updatedAt
+                                type
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    """
+
+    class Type(Enum):
+        CUSTOMERS = "customers"
+        ORDERS = "orders"
+        DRAFT_ORDERS = "draftOrders"
+        PRODUCTS = "products"
+        PRODUCT_IMAGES = ["products", "images"]
+        PRODUCT_VARIANTS = "productVariants"
+        COLLECTIONS = "collections"
+        LOCATIONS = "locations"
+
+    sort_key = "UPDATED_AT"
+    record_composition = {"new_record": "Metafield"}
+
+    metafield_fields: List[Field] = [
+        "__typename",
+        "id",
+        "namespace",
+        "value",
+        "key",
+        "description",
+        "createdAt",
+        "updatedAt",
+        "type",
+    ]
+
+    @property
+    def query_name(self) -> str:
+        if isinstance(self.type, list):
+            return self.type[0]
+        elif isinstance(self.type, str):
+            return self.type
+        else:
+            raise Exception(f"Invalid type for `query_name`: {self.type}.")
+
+    @property
+    @abstractmethod
+    def type(self) -> Type:
+        """
+        Defines the Metafield type to fetch.
+        """
+
+    def get_edge_node(self, name: str, fields: Union[List[str], List[Field], str]) -> Field:
+        """
+        Defines the edge of the graph and it's fields to select for Shopify BULK Operaion.
+        https://shopify.dev/docs/api/usage/bulk-operations/queries#the-jsonl-data-format
+        """
+        return Field(name=name, fields=[Field(name="edges", fields=[Field(name="node", fields=fields)])])
+
+    @property
+    def query_nodes(self) -> List[Field]:
+        """
+        List of available fields:
+        https://shopify.dev/docs/api/admin-graphql/unstable/objects/Metafield
+        """
+        # define metafield node
+        metafield_node = self.get_edge_node("metafields", self.metafield_fields)
+
+        if isinstance(self.type, list):
+            return ["__typename", "id", self.get_edge_node(self.type[1], ["__typename", "id", metafield_node])]
+        elif isinstance(self.type, str):
+            return ["__typename", "id", metafield_node]
+        else:
+            raise Exception(f"Invalid type for `query_nodes`: {self.type}.")
+
+    def record_process_components(self, record: Optional[Mapping[str, Any]] = None) -> Optional[Iterable[Mapping[str, Any]]]:
+        if record:
+            # resolve parent id from `str` to `int`
+            record["owner_id"] = self.tools.resolve_str_id(record.get(BULK_PARENT_KEY))
```

Comment:
Yes, it's possible for a `parent` record. They typically don't have a `__parentId` field inside,  but the `record component` would have the `__parentId`. Since we read each line, and suddenly could face the line (a record) without `__parentId` we will have ValueError. I prefer not to stop the sync in  this case, but rather return `None`